### PR TITLE
feat(tools): add eslint/prettier

### PR DESCRIPTION
**Pull request by nhcarrigan**
Checklist:

- [X] I have read [freeCodeCamp's contribution guidelines](https://contribute.freecodecamp.org).
- [X] My pull request has a [descriptive title](https://contribute.freecodecamp.org/#/how-to-open-a-pull-request?id=prepare-a-good-pr-title) (**not** a vague title like `Update index.md`)

Closes #XXXXX

The fact that we don't have linting/formatting configured is driving me nuts.

Opening this up as a draft so we can discuss rules/plugins - as it stands now:

image

### CODE REVIEW

File: .eslintrc.json [hunk 1]
```diff
@@ -0,0 +1,22 @@
+{
+  "env": {
+    "es6": true,
```
Comment (ojeytonwilliams) [inline .eslintrc.json:3]:
```suggestion
    "es2021": true,
```
The parser couldn't handle some of the new(ish) JS syntax, but this seems to fix the majority of those errors.

File: .prettierrc [hunk 2]
```diff
@@ -0,0 +1,9 @@
+{
+  "endOfLine": "lf",
+  "semi": true,
+  "singleQuote": true,
+  "jsxSingleQuote": true,
+  "tabWidth": 2,
+  "trailingComma": "none",
+  "arrowParens": "avoid"
+}
```
Comment (ojeytonwilliams) [inline .prettierrc:9]: Do we need to customise prettier? I know I'm the one that tends to over-configure things, but can we leave this empty?

**Review by ojeytonwilliams — COMMENTED** [review]
I'm all in favour of this.  Things I would suggest:

- Adding a format script
- Adding a CI lint + prettier check

I can handle that and help with fixing the errors, if you like.  Let me know how you'd like to proceed.

**Comment by ojeytonwilliams** [comment]
Thinking about it, since prettier will be very noisy, how about we get all that done separately and then apply the linting separately?

File: .prettierrc (hunk 2 above)
Comment (nhcarrigan) [reply .prettierrc:9]: FWIW I pulled this from our /learn repo - figured keeping the style matching as much as possible was a good idea.

**Comment by nhcarrigan** [comment]
> Adding a format script
> Adding a CI lint + prettier check

These are on the list, but wanted to get the rules in place first 🙂

File: .prettierrc (hunk 2 above)
Comment (ojeytonwilliams) [reply .prettierrc:9]: Okay, that's a good enough reason for me.

**Comment by nhcarrigan** [comment]
Okay @ojeytonwilliams I broke them down into separate commits to facilitate review. All of the lint errors should be fixed.

A couple of notes:
- I added the boilerplate infrastructure files to the eslintignore - there are some linting errors in there but the code is too delicate to mess with, so best to ignore them I think.
- The secure multiplayer game uses .mjs files which did strange things to the ESLINT for the tests - I had to configure the parser specially for that one test file.

File: .eslintrc.json [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+    "$": true,
+    "ga": true,
+    "jQuery": true,
+    "router": true
+  },
+  "settings": {
+    "react": {
+      "version": "16.4.2"
+    }
+  },
```
Comment (ojeytonwilliams) [inline .eslintrc.json:20]:
```suggestion
```
Let's move this to the override

File: .eslintrc.json [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+  "extends": ["eslint:recommended", "plugin:react/recommended", "prettier"],
```
Comment (ojeytonwilliams) [inline .eslintrc.json:21]:
```suggestion
  "extends": ["eslint:recommended", "prettier"],
```
this too

File: .eslintrc.json [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+  "overrides": [
+    {
+      "files": ["apps/secure-real-time-multiplayer-game/tests/1_unit-tests.js"],
+      "parserOptions": {
+        "sourceType": "module"
+      }
+    }
```
Comment (ojeytonwilliams) [inline .eslintrc.json:28]:
```suggestion
    },
    {
      "files": ["**/*.jsx"],
      "settings": {
        "react": {
          "version": "15.7.0"
        }
      },
      "extends": ["plugin:react/recommended"],
      "parserOptions": {
        "ecmaFeatures": {
          "jsx": true
        }
      }
    }
```
jsx files were getting ignored, but we can force eslint to pay attention via the `files` override. Also, not that it makes much difference, this is the React version we're using.

File: apps/american-british-translator/components/translator.js [hunk 6]
```diff
@@ -35,14 +35,14 @@ class Translator {
 
     // Title Replacement
     for([american, british] of Object.entries(americanToBritishTitles) ) {
-      american = american.replace('.', '\.');
+      american = american.replace('.', '\\.');
```
Comment (ojeytonwilliams) [inline apps/american-british-translator/components/translator.js:38]: Nice catch!  I hadn't realised that `mrr` translated to `Mr`.

**Review by ojeytonwilliams — CHANGES_REQUESTED** [review]
I can fix the new errors this turns up, but didn't want to push anything until you'd had a chance to look.

**Comment by nhcarrigan** [comment]
image

Not too bad.

**Review by ojeytonwilliams — APPROVED** [review]
LGTM :+1:

**Pull request merged**